### PR TITLE
Do not steal libbpf's file descriptors

### PR DESCRIPTION
**Pull request by GermanCoding**
This is for I/O (and possibly memory) safety: We currently create some `mmap` on a file descriptor that we do *not* own - it is owned by libbpf. Generally, the FD lifetime is bound by the `Map`'s lifetime. However, our `RingBuffer` struct is not bound to the Map's lifetime, and may well outlive it. This is unsafe, because if the `Map` gets dropped, the FD is lost and then we rely on things that may no longer be there (i.e. there's no guarantee that our mmap still exists).

This change fixes this, by duplicating the file descriptor. Thus, we actually own it and can therefore be sure of its lifetime. This makes our RingBuffer truly independent of the original `Map` reference lifetime, as implied by the function contract of `RingBuffer::new()`.

### CODE REVIEW

**Comment by fujita** [comment]
Thanks!

**Comment by GermanCoding** [comment]
A small correction to my original post above:

The `mmap` manual page says:

> After the mmap() call has returned, the file descriptor, fd, can
    be closed immediately without invalidating the mapping.

So my statement that the mmap might go away is wrong - this doesn't happen, sorry for the mistake.

In any case, the basic idea of this PR still holds: By duplicating the FD we ensure that the lifetime of our ring buffer is independent of the map lifetime - otherwise, even if the mmap is still there, it won't be updated anymore. Also, I've no idea how tokio reacts to epolling on a raw fd that has been closed. This is just the safer choice IMHO.

TL;DR: I still consider this PR to be correct, even though my initial argumentation was slightly off.

**Comment by fujita** [comment]
Ah, mmap() internally holds the file descriptor. So even if the Map object is dropped, the file isn't closed, I guess. Needs an experiment but we could revert this change.

**Comment by GermanCoding** [comment]
Yeah, looks like I wasn't aware of the inner workings of mmap - should have checked this beforehand, sorry. That's what happens if you make assumptions without checking them 😄.

I did just test this: Even if you fully drop the map object, the map stays alive as long as you have the mmaps open.

**Comment by fujita** [comment]
Thanks for the confirmation!
I'll revert the commit. Things like this always happen so no problem for me :smile:
